### ISSUE TITLE
Design: Explicit state transition handling in evaluator

### ISSUE THREAD

**Issue by sondreso** [equinor/ert#1691]
At the moment we have little to no validation of the state transitions that the incoming events represent. There are a lot of transitions that we do not allow, and many of them could occur if events would come out of order. Currently this is mitigated by the fact that most events for each entity (eg. step or job) in our status model is reported by the same agent, and along with the properties of websockets, this means that they will not come out of order. While this holds true for events for the same entity, we have noticed during development that ordering of events ensemble level events and the step/job level events might not always follow the assumptions we have made. For example: We should be able to handle the following ordering: `EnsembleStart` -> `EnsembleStop` -> `StepUpdate`. There are several steps that can be taken independently to improve our handling of transitions in status updates: 

1. Make the evaluator able to handle ensemble level events that are out of order wrt. step/job events.
2. Move state handling into the ensemble object itself (base ensemble)
2. Convert all steps and jobs into state machines, such that for example the following order of events would not put the ensemble into an inconsistent state ((x) represents the order they were sent in): `Waiting(1)` -> `Finished(3)` -> `Running(2)`. Currently this would result in the step/job ending up in the running state, even though this does not represent the actual state of the entity. 

Depends on: #1690

**Comment by frode-aarstad** [comment]
This issue should result in implementation issue(s)

**Comment by DanSava** [comment]
After a bit of design prototyping, I propose the following:

1. Collapsing all possible states to only 4: 
-  Wating 
- Running
- Finished
- Success
 with the following state transitions 
![Screenshot from 2021-06-09 11-09-20](https://user-images.githubusercontent.com/4508053/121321622-f0b83680-c916-11eb-9121-3d7a50524d79.png)
In the case of an ensemble with one realization one steps and 2 jobs the transitions should look something like this:
![Screenshot from 2021-06-09 11-37-51](https://user-images.githubusercontent.com/4508053/121321861-265d1f80-c917-11eb-81d8-860b5eef3cbf.png)
2. Only sending update messages for the jobs so that means stop sending step update messages. This will avoid the issue of having an inconsistent state between the step and the children jobs possible in the case of messages out of order.
3. Only update parent state using the current children states for Realizations and Steps.

**Comment by jondequinor** [comment]
> 2. Only sending update messages for the jobs so that means stop sending step update messages. This will avoid the issue of having an inconsistent state between the step and the children jobs possible in the case of messages out of order.

A consequence of this is that you will either a) have an implicit job doing the work (or just reporting) of the step; or b) arbitrarily say that the first/last job should report the work of the step.

The step work here is defined as handling I/O. For legacy it additionally means queue-level idiosyncrasies, for Prefect it means tasks.

It sure is _easier_, but not _simpler_.

**Comment by jondequinor** [comment]
> 3. Only update parent state using the current children states for Realizations and Steps.

I think that's how it is today for Realization. Not so for step, see above comment as to why.

**Comment by DanSava** [comment]
> A consequence of this is that you will either a) have an implicit job doing the work (or just reporting) of the step; or b) arbitrarily say that the first/last job should report the work of the step.

Having an implicit job that would be doing the work of the step would work, but that would just make things confusing, same is true if you have to do something special for the first of last job.

If the step has to do any meaningful work, and it does as you have pointed out, then the step needs to send it's own updates :disappointed:

**Comment by jondequinor** [comment]
I think the question boils down to: how do we want to model our ensemble state? I can think of three useful abstractions:

- The current "bottom up" approach which ultimately will [result in code riddled with a disproportionate number of convoluted conditional branches.](https://www.cis.upenn.edu/~lee/06cse480/lec-HSM.pdf) Example [1](https://github.com/equinor/ert/blob/main/res/job_queue/job_queue_node.py), [2](https://github.com/equinor/ert/blob/main/ert_shared/ensemble_evaluator/entity/snapshot.py#L138). The latter of which I am co-author. It's a great start, but it just doesn't scale/read/age well.
- Task switching using _[behavioral trees](https://arxiv.org/format/1709.00084)_ (BT).
- _Finite state machines_ (SM). I don't think we can model the experiment state without a pretty nasty state explosion.
- Extended SM. Extended either with concurrency (which looks like the second illustration in https://github.com/equinor/ert/issues/1691#issuecomment-857506467); or extended with a hierarchy.

BTs are likely more straight forward—conceptually—than SMs. BTs are certainly more popular. They also closely resembles our current snapshot structure, but in order to become a BT we would have to populate the snapshot with _control nodes_ (we can regard the current nodes, or keys in the snapshot dict, as _execution nodes_).

Hierarchical SMs (HSM) use behavioral inheritance to vastly reduce the complexity of modelling state in complex systems. An example of this is the `Waiting` state of `J1`. In the proposed visualization, this is its own state (as I interpret it). With behavioral inheritance, this state is a substate of `S1:Waiting`, `R1:Waiting` and `Ens:Waiting`. A transition of `J1:Waiting->Running` could actually change not only the job's state, but all its parents too. But more importantly, a subsequent transition `J2:Waiting->Running` wouldn't necessarily affect the parent states,only `J2`'s.

HSM employs the same principles as inheritance in OOP, so it's perhaps more familiar. But the distance from where we are today, to BTs, seems smaller.

@sondreso most likely has experience with BTs: what do you think?

**Comment by jondequinor** [comment]
I made an attempt at implementing this with a basic hierarchical state machine. By basic I mean without the extensions defined by statecharts (conditions on transitions, entry/exit actions, etc).

A hierarchical state machine is a state machine that allows a state to be a state machine.

With the basic HSM, I modeled a small portion of an ensemble (one step with two jobs) where the running state of the step was a state machine that encapsulated the states of the two jobs. So the state of the step at some point becomes a state machine (running), that itself is driven forward by events. The running step state machine is made up of two job state machines.

![Untitled Diagram(4)](https://user-images.githubusercontent.com/53553193/122452422-87d15e00-cfa9-11eb-8b89-9915004567a5.png)


Events are dispatched to the topmost state machine (the step) and propagated to the lower levels. This is contrary to the ultimate hook pattern, where events are supposed to propagate towards the "higher levels" if not handled by the lower levels. But that leads to complex dispatching in our case, because each realization will be disjoint from each other. So if an event comes in, you'll need to locate e.g. the correct job in the correct realization. You have a bazillion leaf nodes, which isn't common it seems.

By introducing multiple realizations, we also get a concurrent state machine, since the state of any realization can transition independently.

It is suggested by someʷʰᵒ that you use OOP to make implementing HSM smoother. This way, the code could be modularized into ens, real, step, job; as well as having the various implementations encapsulated. Imagine an API like this:

```python
ens_state_machine = EnsembleStateMachine(ensemble)
…
ens_state_machine.dispatch(cloudevent)
```

Where `EnsembleStateMachine` is

```python
class EnsembleStateMachine(StateMachine):
    def __init__(self, ensemble: _Ensemble) -> None:
        for real in ensemble.get_reals().values():
            self.add_substate(RealizationStateMachine(real))
            # add transitions here
```

Still, it's quite complex. Maybe I've misunderstood something.

Anyway, I will look into a simple behavioral tree implementation of the same thing.

**Comment by jondequinor** [comment]
Having looked more into behavioral trees (BT), and thinking about what we're actually trying to achieve, I've concluded that BTs are too much geared towards dynamic behavior of agents. Our case is merely about reacting to external events. We _might_ perform some action in the future (restart, timeout, etc), but future proofing that now using BTs does not seem feasible.

So I revisited the hierarchical state machine idea, but without sticking to any formalism. So I made an implementation of flat state machines in a acyclic graph.

Code here https://gist.github.com/jondequinor/aaf3fee04e406c880b5387c28f469a85

Events flow through the graph from the root, which represents the ensemble, and down towards the leaves—the jobs. Transitions are propagated up the tree, but only to the parent. So a transition in jobs will only be propagated to the step, and so on.

It then becomes trivial to reason about the state of the jobs in the step; about state of steps in reals, etc.

In the code, two cases of data inconsistency are specifically handled:
        ("/0", "ENS_START"),
        ("/0/r1/s1", "STEP_STARTED"),
        ("/0/r1/s1/j1", "JOB_STARTED"),
        ("/0/r1/s1/j1", "JOB_RUNNING"),
        ("/0/r1/s1/j2", "JOB_SUCCESS"),
        ("/0/r1/s1", "STEP_SUCCESS"),
        ("/0/r2/s2", "STEP_STARTED"),
        ("/0/r2/s2/j1", "JOB_STARTED"),
        ("/0/r2/s2/j1", "JOB_RUNNING"),
        ("/0/r2/s2/j1", "JOB_SUCCESS"),
        ("/0/r2/s2/j2", "JOB_STARTED"),
        ("/0/r2/s2/j2", "JOB_SUCCESS"),
        ("/0/r2/s2/j2", "JOB_RUNNING"),
        ("/0/r2/s2", "STEP_SUCCESS"),
        ("/0", "ENS_SUCCESS"),


1. `/0/r1/s1/j1` will be running, then the next event is `/0/r1/s1/j2`'s `SUCCESS` event. This is handle by marking (not transitioning) all jobs before it as `SUCCESS`.
2. `/0/r2/s2/j2` will receive an event that indicates it should go from `SUCCESS` to `RUNNING`. This transition is simply ignored.
